ddl/ddl_api.go

```diff
@@ -4942,6 +4942,27 @@ func (d *ddl) AlterTableSetTiFlashMode(ctx sessionctx.Context, ident ast.Ident,
 		return fmt.Errorf("unsupported TiFlash mode %s", mode)
 	}
 
+	// Ban setting tiflash mode for tables in system database.
+	if util.IsMemOrSysDB(schema.Name.L) {
+		return errors.Trace(dbterror.ErrUnsupportedAlterTiFlashModeForSysTable)
+	} else if tb.Meta().TempTableType != model.TempTableNone {
+		return dbterror.ErrOptOnTemporaryTable.GenWithStackByArgs("set tiflash mode")
+	}
+
+	// Ban setting tiflash mode for tables which has charset not supported by TiFlash
+	for _, col := range tb.Cols() {
+		_, ok := charset.TiFlashSupportedCharsets[col.GetCharset()]
+		if !ok {
+			return dbterror.ErrAlterTiFlashModeForUnsupportedCharsetTable.GenWithStackByArgs(col.GetCharset())
+		}
+	}
```

Comment:
Is it possible to extract common functions for these logic? Seems that these logic can be reused not only for SET TIFLASH MODE, but also for all TIFLASH related operations.